pkg/operator/observe_config.go

```diff
@@ -93,90 +107,162 @@ func NewConfigObserver(
 	return c
 }
 
-func observeKubeAPIServerPublicInfo(listers Listers, observedConfig map[string]interface{}) (map[string]interface{}, error) {
-	kubeAPIServerPublicInfo, err := listers.configmapLister.ConfigMaps(kubeAPIServerNamespaceName).Get("public-info")
-	if err != nil && !errors.IsNotFound(err) {
-		return nil, err
+func observeDefaultNodeSelector(listers Listers, currentConfig map[string]interface{}) (observedConfig map[string]interface{}, errs []error) {
+	observedConfig = map[string]interface{}{}
+	defaultNodeSelectorPath := []string{"projectConfig", "defaultNodeSelector"}
+	currentDefaultNodeSelector, _, _ := unstructured.NestedString(currentConfig, defaultNodeSelectorPath...)
+
+	publicInfo, err := listers.configmapLister.ConfigMaps(kubeAPIServerNamespaceName).Get("public-info")
+	if errors.IsNotFound(err) {
+		glog.Warningf("configmap/public-info.%v: not found", kubeAPIServerNamespaceName)
+		return
 	}
-	if kubeAPIServerPublicInfo != nil {
-		if val, ok := kubeAPIServerPublicInfo.Data["projectConfig.defaultNodeSelector"]; ok {
-			unstructured.SetNestedField(observedConfig, val, "projectConfig", "defaultNodeSelector")
+	if err != nil {
+		errs = append(errs, fmt.Errorf("configmap/public-info.%v: %v", kubeAPIServerNamespaceName, err))
+		if len(currentDefaultNodeSelector) > 0 {
+			unstructured.SetNestedField(observedConfig, currentDefaultNodeSelector, defaultNodeSelectorPath...)
 		}
+		return
 	}
-
-	return observedConfig, nil
+	defaultNodeSelector := publicInfo.Data["projectConfig.defaultNodeSelector"]
+	if len(defaultNodeSelector) > 0 {
+		unstructured.SetNestedField(observedConfig, defaultNodeSelector, defaultNodeSelectorPath...)
+	}
+	return
 }
 
-// observeEtcdEndpoints reads the etcd endpoints from the endpoints object and then manually pull out the hostnames to
-// get the etcd urls for our config. Setting them observed config causes the normal reconciliation loop to run
-func observeEtcdEndpoints(listers Listers, observedConfig map[string]interface{}) (map[string]interface{}, error) {
-	etcdURLs := []string{}
-	etcdEndpoints, err := listers.endpointLister.Endpoints(etcdNamespaceName).Get("etcd")
-	if err != nil && !errors.IsNotFound(err) {
-		return nil, err
+// observeStorageURLs observes the storage config URLs. If there is a problem observing the current storage config URLs,
+// then the previously observed storage config URLs will be re-used.
+func observeStorageURLs(listers Listers, currentConfig map[string]interface{}) (observedConfig map[string]interface{}, errs []error) {
```

Comment:
instead of this, let's see if we can just use the "normal" etcd service that was introduced last week.